nvm.sh

```diff
@@ -191,6 +191,71 @@ nvm_find_up() {
   nvm_echo "${path}"
 }
 
+nvm_find_package_json() {
+  local dir
+  dir="$(nvm_find_up 'package.json')"
+  if [ -e "${dir}/package.json" ]; then
+    nvm_echo "${dir}/package.json"
+  fi
+}
+
+nvm_get_node_from_pkg_json() {
+  local engines_object
+  engines_object=''
+  local open_brackets
+  open_brackets=0
+  local closed_brackets
+  closed_brackets=0
+  local in_quotes
+  in_quotes='false'
+
+  command printf "%s" "$1" \
+    | command tr -d '[[:space:]]' \
+    | command grep -o '"engines":{".*' \
+    | command cut -c 11- \
+    | command grep -o . \
+    | while read -r i; do
+        engines_object="$engines_object$i"
+        if [ "$i" = '"' ]; then
+          if [ "$in_quotes" = 'false' ]; then
+            in_quotes='true'
+         else
+            in_quotes='false'
+         fi
+        fi
+        if [ "$in_quotes" = 'false' ]; then
+          if [ "$i" = '{' ]; then
+            open_brackets=$((open_brackets+1))
+          elif [ "$i" = '}' ]; then
+            closed_brackets=$((closed_brackets+1))
+          fi
+        fi
+        if [ "$open_brackets" -ne 0 ] && [ "$open_brackets" -eq "$closed_brackets" ]; then
+          command printf "%s" "$engines_object" \
+            | command grep -o '"node":"v\?[0-9.]\+"' \
+            | command tr -d '"' \
+            | command awk -F: '{ print $2 }'
+          return 0
+        fi
+    done
+    return 2
+}
+
+nvm_package_json_version() {
+  export PKG_JSON_VERSION=''
+  local pkg_json_path=''
```

Comment:
the `local` declaration and the initial assignment must happen on separate lines, due to ksh